contracts/modules/VersionManager.sol

```diff
@@ -0,0 +1,276 @@
+
+            // Store static call information to optimise its use by wallets
+            bytes4[] memory sigs = IFeature(_features[i]).getStaticCallSignatures();
+            for(uint256 j = 0; j < sigs.length; j++) {
+                staticCallSignatures[newVersion].push(sigs[j]);
+                staticCallExecutors[newVersion][sigs[j]] = _features[i];
+            }
+        }
+        featuresToInit[newVersion] = _featuresToInit;
+        
+        emit VersionAdded(newVersion);
+    }
+
+    /* ***************** View Methods ************************* */
+
+    /**
+     * @inheritdoc IVersionManager
+     */
+    function isFeatureAuthorised(address _wallet, address _feature) public view override returns (bool) {
+        return isFeatureInVersion[_feature][walletVersions[_wallet]] || _feature == address(this);
+    }
+
+    /**
+     * @inheritdoc IFeature
+     */
+    function getRequiredSignatures(address /* _wallet */, bytes calldata _data) external view override returns (uint256, OwnerSignature) {
+        bytes4 methodId = Utils.functionPrefix(_data);
+        // This require ensures that the RelayerManager cannot be used to call a featureOnly VersionManager method
+        // that calls a Storage or the BaseWallet for backward-compatibility reason
+        require(methodId == UPGRADE_WALLET_PREFIX || methodId == ADD_MODULE_PREFIX, "VM: unknown method");     
+        return (1, OwnerSignature.Required);
+    }
+
+    /**
+     * @notice This method delegates the static call to a target feature
+     */
+    fallback() external {
+        uint256 version = walletVersions[msg.sender];
+        address feature = staticCallExecutors[version][msg.sig];
+        require(feature != address(0), "VM: Static call not supported for wallet version");
+
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            calldatacopy(0, 0, calldatasize())
+            let result := staticcall(gas(), feature, 0, calldatasize(), 0, 0)
+            returndatacopy(0, 0, returndatasize())
+            switch result
+            case 0 {revert(0, returndatasize())}
+            default {return (0, returndatasize())}
+        }
+    }
+
+    /* ***************** Wallet Upgrade ************************* */
+
+    /**
+     * @inheritdoc IFeature
+     */
+    function init(address _wallet) public override(IModule, BaseFeature) onlyWallet(_wallet) {
+        doUpgradeWallet(_wallet, featuresToInit[lastVersion]);
+    }
+
+    /**
+     * @notice Upgrade a wallet to the latest version.
+     * @dev It's cheaper to pass features to init as calldata than reading them from storage
+     * @param _wallet the wallet to upgrrade
+     * @param _featuresToInit the subset of features that need to be initialized for the walleet
+     */
+    function upgradeWallet(address _wallet, address[] calldata _featuresToInit) external onlyWalletOwnerOrFeature(_wallet) onlyWhenUnlocked(_wallet) {
+        doUpgradeWallet(_wallet, _featuresToInit);
```

Comment:
I understand that we pass `_featuresToInit` in calldata to avoid a storage read and reduce the gas cost but I think it is error prone. If you pass e.g. an incomplete list of features to init then the wallet will be in the target version but in an inconsistent state (i.e. no all features initialised). It is much safer to load the list of features from `featuresToInit` since we already check in `doUpgradeWallet` which ones need to be initialised. Given that a SLOAD is only 200 gas/word I don't think the cost saving is worth the fragility.

Reply:
An SLOAD costs 800 gas since Istanbul and is likely to increase even more (see https://notes.ethereum.org/@vbuterin/BkrNbeAfD). A middle ground solution could be to check that the calldata array is the same size as the storage array?

Reply:
Or we could store a hash of the `featuresToInit` array in `addVersion` and compare this hash to the hash of the calldata `_featuresToInit` in `upgradeWallet()` to at least compress the cost to a single SLOAD regardless of the number of features.

Reply:
Anyway for now I'll just remove that calldata parameter to keep things simple.